--- a/exp_chan.c
+++ b/exp_chan.c
@@ -51,6 +51,8 @@ static void		ExpWatchProc _ANSI_ARGS_((C
 		            int mask));
 static int		ExpGetHandleProc _ANSI_ARGS_((ClientData instanceData,
 		            int direction, ClientData *handlePtr));
+void			exp_background_channelhandler _ANSI_ARGS_((ClientData,
+		            int));
 
 /*
  * This structure describes the channel type structure for Expect-based IO:
@@ -58,7 +60,7 @@ static int		ExpGetHandleProc _ANSI_ARGS_
 
 Tcl_ChannelType expChannelType = {
     "exp",				/* Type name. */
-    ExpBlockModeProc,			/* Set blocking/nonblocking mode.*/
+    TCL_CHANNEL_VERSION_2,
     ExpCloseProc,			/* Close proc. */
     ExpInputProc,			/* Input proc. */
     ExpOutputProc,			/* Output proc. */
@@ -68,6 +70,7 @@ Tcl_ChannelType expChannelType = {
     ExpWatchProc,			/* Initialize notifier. */
     ExpGetHandleProc,			/* Get OS handles out of channel. */
     NULL,				/* Close2 proc */
+    ExpBlockModeProc,			/* Set blocking/nonblocking mode.*/
 };
 
 typedef struct ThreadSpecificData {
--- a/exp_clib.c
+++ b/exp_clib.c
@@ -37,6 +37,14 @@ would appreciate credit if this program
 # endif
 #endif
 
+#ifdef HAVE_UNISTD_H
+# include <unistd.h>
+#endif
+
+//#ifdef HAVE_SYS_WAIT_H
+# include <sys/wait.h>
+//#endif
+
 #ifdef HAVE_SYS_FCNTL_H
 #  include <sys/fcntl.h>
 #else
@@ -1727,13 +1735,14 @@ int exp_logfile_all = FALSE;	/* if TRUE,
 int exp_loguser = TRUE;		/* if TRUE, user sees interactions on stdout */
 
 
-char *exp_printify();
+char *exp_printify(char *);
 int exp_getptymaster();
-int exp_getptyslave();
+int exp_getptyslave(int, int, CONST char *);
 
 #define sysreturn(x)	return(errno = x, -1)
 
 void exp_init_pty();
+void exp_init_tty();
 
 /*
    The following functions are linked from the Tcl library.  They
@@ -2253,6 +2262,7 @@ exp_spawnl TCL_VARARGS_DEF(char *,arg1)
 		argv[i] = va_arg(args,char *);
 		if (!argv[i]) break;
 	}
+	va_end(args);
 	i = exp_spawnv(argv[0],argv+1);
 	free((char *)argv);
 	return(i);
@@ -2726,6 +2736,7 @@ exp_expectl TCL_VARARGS_DEF(int,arg1)
 		/* Ultrix 4.2 compiler refuses enumerations comparison!? */
 		if ((int)type < 0 || (int)type >= (int)exp_bogus) {
 			fprintf(stderr,"bad type (set %d) in exp_expectl\n",i);
+			va_end(args);
 			sysreturn(EINVAL);
 		}
 
@@ -2791,6 +2802,7 @@ exp_fexpectl TCL_VARARGS_DEF(FILE *,arg1
 		/* Ultrix 4.2 compiler refuses enumerations comparison!? */
 		if ((int)type < 0 || (int)type >= (int)exp_bogus) {
 			fprintf(stderr,"bad type (set %d) in exp_expectl\n",i);
+			va_end(args);
 			sysreturn(EINVAL);
 		}
 
--- a/exp_command.h
+++ b/exp_command.h
@@ -222,8 +222,8 @@ EXTERN int		exp_trap_off _ANSI_ARGS_((ch
 
 EXTERN void		exp_strftime(char *format, const struct tm *timeptr,Tcl_DString *dstring);
 
-#define exp_deleteProc (void (*)())0
-#define exp_deleteObjProc (void (*)())0
+#define exp_deleteProc NULL
+#define exp_deleteObjProc NULL
 
 EXTERN int expect_key;
 EXTERN int exp_configure_count;	/* # of times descriptors have been closed */
@@ -321,7 +321,7 @@ EXTERN void		exp_init_expect_cmds _ANSI_
 EXTERN void		exp_init_most_cmds _ANSI_ARGS_((Tcl_Interp *));
 EXTERN void		exp_init_trap_cmds _ANSI_ARGS_((Tcl_Interp *));
 EXTERN void		exp_init_interact_cmds _ANSI_ARGS_((Tcl_Interp *));
-EXTERN void		exp_init_tty_cmds();
+EXTERN void		exp_init_tty_cmds _ANSI_ARGS_((Tcl_Interp *));
 
 EXTERN ExpState *	expStateCheck _ANSI_ARGS_((Tcl_Interp *,ExpState *,int,int,char *));
 EXTERN ExpState *       expStateCurrent _ANSI_ARGS_((Tcl_Interp *,int,int,int));
--- a/exp_log.c
+++ b/exp_log.c
@@ -174,7 +174,10 @@ expStdoutLog TCL_VARARGS_DEF(int,arg1)
     force_stdout = TCL_VARARGS_START(int,arg1,args);
     fmt = va_arg(args,char *);
 
-    if ((!tsdPtr->logUser) && (!force_stdout) && (!tsdPtr->logAll)) return;
+    if ((!tsdPtr->logUser) && (!force_stdout) && (!tsdPtr->logAll)) {
+	va_end(args);
+	return;
+    }
 
     (void) vsprintf(bigbuf,fmt,args);
     expDiagWriteBytes(bigbuf,-1);
--- a/exp_main_sub.c
+++ b/exp_main_sub.c
@@ -49,14 +49,15 @@ char exp_version[] = PACKAGE_VERSION;
 #define NEED_TCL_MINOR		5
 
 char *exp_argv0 = "this program";	/* default program name */
-void (*exp_app_exit)() = 0;
-void (*exp_event_exit)() = 0;
+void (*exp_app_exit) _ANSI_ARGS_((Tcl_Interp *)) = 0;
+void (*exp_event_exit) _ANSI_ARGS_((Tcl_Interp *)) = 0;
 FILE *exp_cmdfile = 0;
 char *exp_cmdfilename = 0;
 int exp_cmdlinecmds = FALSE;
 int exp_interactive =  FALSE;
 int exp_buffer_command_input = FALSE;/* read in entire cmdfile at once */
 int exp_fgets();
+int exp_tty_cooked_echo(Tcl_Interp *interp, exp_tty *tty_old, int *was_raw, int *was_echo);
 
 Tcl_Interp *exp_interp;	/* for use by signal handlers who can't figure out */
 			/* the interpreter directly */
@@ -920,7 +921,6 @@ int sys_rc;
 	    char file[200];
 	    char *home;
 	    int fd;
-	    char *getenv();
 
 	    if ((NULL != (home = getenv("DOTDIR"))) ||
 		(NULL != (home = getenv("HOME")))) {
--- a/exp_pty.c
+++ b/exp_pty.c
@@ -134,8 +134,8 @@ int timeout;
 	return(cc);
 }
 
-static RETSIGTYPE (*oldAlarmHandler)();
-static RETSIGTYPE (*oldHupHandler)();
+static RETSIGTYPE (*oldAlarmHandler)(int);
+static RETSIGTYPE (*oldHupHandler)(int);
 static time_t current_time;	/* time when testing began */
 
 /* if TRUE, begin testing, else end testing */
--- a/exp_trap.c
+++ b/exp_trap.c
@@ -53,7 +53,8 @@ static struct trap {
 
 int sigchld_count = 0;	/* # of sigchlds caught but not yet processed */
 
-static int eval_trap_action();
+static int
+eval_trap_action(Tcl_Interp *, int, struct trap *, int);
 
 static int got_sig;		/* this records the last signal received */
 				/* it is only a hint and can be wiped out */
--- a/exp_tty.c
+++ b/exp_tty.c
@@ -572,7 +572,7 @@ Exp_SystemCmd(
     char **argv)
 {
 	int result = TCL_OK;
-	RETSIGTYPE (*old)();	/* save old sigalarm handler */
+	RETSIGTYPE (*old)(int);	/* save old sigalarm handler */
 #define MAX_ARGLIST 10240
 	int i;
 
--- a/exp_win.h
+++ b/exp_win.h
@@ -8,8 +8,8 @@ would appreciate credit if you use this
 
 #include <tcl.h> /* For _ANSI_ARGS_ */
 
-int exp_window_size_set();
-int exp_window_size_get();
+int exp_window_size_set _ANSI_ARGS_ ((int));
+int exp_window_size_get _ANSI_ARGS_ ((int));
 
 void  exp_win_rows_set    _ANSI_ARGS_ ((char* rows));
 char* exp_win_rows_get    _ANSI_ARGS_ ((void));
--- a/pty_termios.c
+++ b/pty_termios.c
@@ -105,6 +105,7 @@ with openpty which supports 4000 while p
 
 void expDiagLog();
 void expDiagLogPtr();
+char *expErrnoMsg(int errorNo);
 
 #include <errno.h>
 /*extern char *sys_errlist[];*/
@@ -189,6 +190,7 @@ static char slave_name[MAXPTYNAMELEN];
 #endif /* HAVE_SCO_CLIST_PTYS */
 
 #ifdef HAVE_OPENPTY
+#include <pty.h>
 static char master_name[64];
 static char slave_name[64];
 #endif
@@ -204,7 +206,7 @@ char *name;		/* name of pty */
 {
 #define MAX_ARGLIST 10240
 	char buf[MAX_ARGLIST];	/* overkill is easier */
-	RETSIGTYPE (*old)();	/* save old sigalarm handler */
+	RETSIGTYPE (*old)(int);	/* save old sigalarm handler */
 	int pid;
 	
 	old = signal(SIGCHLD, SIG_DFL);
@@ -264,7 +266,7 @@ char *name;		/* name of pty */
 {
 #define MAX_ARGLIST 10240
 	char buf[MAX_ARGLIST];	/* overkill is easier */
-	RETSIGTYPE (*old)();	/* save old sigalarm handler */
+	RETSIGTYPE (*old)(int);	/* save old sigalarm handler */
 
 #ifdef STTY_READS_STDOUT
 	sprintf(buf,"%s %s > %s",STTY_BIN,s,name);
